### ISSUE TITLE
Static field initialization

### ISSUE THREAD

**Issue by bernd5** [dotnet/csharplang#2893]
According to "§17.4.5.1 static field initialization" static field initializers are executed after static default initialization followed by an optional static constructor execution.
It is garanteed that all initializers are executed before the static constructor gets executed and that static initialization must be done before first use.

The second sample in this section says that the output must be 
```Init B
8 Init A
9 1 1
```

My question:
Is it garanteed by the spec that static initialization is deferred to the first usage of a member? Is it not allowed to do static intialization for all types before executing Main?
If it is so, where it is defined in the spec?

**Comment by YairHalberstadt** [comment]
It is not defined when static initialisation will take place.

Indeed different runtimes do this at different times. The general trend has been for this to become increasingly lazy over time.

**Comment by bernd5** [comment]
But if so, that sample is just wrong. There is no initalization dependency between A and B.
If static intialization would be executed right before Main the output could be the same as in sample one.

**Comment by YairHalberstadt** [comment]
Can you add a link to that section please?

**Comment by bernd5** [comment]
[static-field-initialization](https://github.com/dotnet/csharplang/blob/892af9016b3317a8fae12d195014dc38ba51cf16/spec/classes.md#static-field-initialization)

**Comment by YairHalberstadt** [comment]
Static initialisation can definitely occur before Main, as this example will prove:

```csharp
using System;
public class C {
    static int b = a + 5;
    static int  a;
    public static void Main() {
        a = 5;
        WriteB();
    }
    
    private static void WriteB() => Console.WriteLine(b);
}
```

This prints 5, not 10

**Comment by YairHalberstadt** [comment]
The reason for that sample is that when there is a static constructor, I think all static field initialisers are moved to the static constructor.

The static constructor is guaranteed to be lazy, unlike static field initialisers.

**Comment by bernd5** [comment]
No, that is no prove. As defined in 17.4.5.1 static field initialization is done in textual order, right after default initialization.

In your sample a and b are default initialized to zero. Than b is initialized (0 + 5 => 5).

**Comment by YairHalberstadt** [comment]
The relevant section of the spec is in static constructors, and is:

> The execution of a static constructor is triggered by the first of the following events to occur within an application domain:
>
> 1. An instance of the class type is created.
> 2. Any of the static members of the class type are referenced.

**Comment by YairHalberstadt** [comment]
> In your sample a and b are default initialized to zero. Than b is initialized (0 + 5 => 5).

That clearly happens before Main, or a would already be 5.

**Comment by bernd5** [comment]
So executing a static constructor before instance creation or static member access is not allowed? Or is it just garanteed that static constructor executions is executed at least before those events?

In the CLR this is not a problem because it can execute static initialization right after JITing. Afterwards no check is required to access static fields.
But for AOT scenarios it would be nice to do all static initialization just before application startup.

**Comment by huoyaoyuan** [comment]
Are you meaning order of multiple static members in 1 class?
That's a bit confusing. An instance field initialization cannot reference to another instance field. Dependencies must be explicitly specified in constructor code. However, static field initialization is allowed to reference others, and dependency field in the same class may be uninitialized.
I do want specification to clearify/guard about this.
